tests/integration/ambient/wasm_test.go

```diff
@@ -0,0 +1,232 @@
+}
+
+func installWasmExtension(ctx framework.TestContext, pluginName, wasmModuleURL, imagePullPolicy, pluginVersion, path string) error {
+	args := map[string]any{
+		"WasmPluginName":     pluginName,
+		"TestWasmModuleURL":  wasmModuleURL,
+		"WasmPluginVersion":  pluginVersion,
+		"TargetAppName":      GetTarget().(echo.Instances).NamespacedName().Name,
+		"TargetGatewayName":  GetTarget().(echo.Instances).ServiceName() + "-gateway",
+		"TargetServiceName":  GetTarget().(echo.Instances).ServiceName(),
+		"TargetWaypointName": constants.DefaultNamespaceWaypoint,
+	}
+
+	if len(imagePullPolicy) != 0 {
+		args["ImagePullPolicy"] = imagePullPolicy
+	}
+
+	if err := applyWasmConfig(ctx, apps.Namespace.Name(), args, path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func uninstallWasmExtension(ctx framework.TestContext, pluginName, path string) error {
+	args := map[string]any{
+		"WasmPluginName": pluginName,
+	}
+	if err := ctx.ConfigIstio().EvalFile(apps.Namespace.Name(), args, path).Delete(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func sendTraffic(ctx framework.TestContext, checker echo.Checker, options ...retry.Option) {
+	ctx.Helper()
+	if len(GetClientInstances()) == 0 {
+		ctx.Fatal("there is no client")
+	}
+	cltInstance := GetClientInstances()[0]
+
+	defaultOptions := []retry.Option{retry.Delay(100 * time.Millisecond), retry.Timeout(200 * time.Second)}
+	httpOpts := echo.CallOptions{
+		To: GetTarget(),
+		Port: echo.Port{
+			Name: "http",
+		},
+		HTTP: echo.HTTP{
+			Path:   "/path",
+			Method: "GET",
+		},
+		Count: 1,
+		Retry: echo.Retry{
+			Options: append(defaultOptions, options...),
+		},
+		Check: checker,
+	}
+
+	_ = cltInstance.CallOrFail(ctx, httpOpts)
+}
+
+func sendTrafficToHostname(ctx framework.TestContext, checker echo.Checker, hostname string, options ...retry.Option) {
+	ctx.Helper()
+	if len(GetClientInstances()) == 0 {
+		ctx.Fatal("there is no client")
+	}
+	cltInstance := GetClientInstances()[0]
+
+	defaultOptions := []retry.Option{retry.Delay(100 * time.Millisecond), retry.Timeout(200 * time.Second)}
+	httpOpts := echo.CallOptions{
+		Address: hostname,
+		Port: echo.Port{
+			Name:        "http",
+			ServicePort: 80,
+			Protocol:    protocol.HTTP,
+		},
+		HTTP: echo.HTTP{
+			Path:    "/path",
+			Method:  "GET",
+			Headers: headers.New().WithHost(fmt.Sprintf("%s.com", GetTarget().ServiceName())).Build(),
+		},
+		Count: 1,
+		Retry: echo.Retry{
+			Options: append(defaultOptions, options...),
+		},
+		Check: checker,
+	}
+
+	_ = cltInstance.CallOrFail(ctx, httpOpts)
+}
+
+// TestTargetRef vs workloadSelector for gateways
+func TestGatewaySelection(t *testing.T) {
+	framework.NewTest(t).
+		Run(func(t framework.TestContext) {
+			crd.DeployGatewayAPIOrSkip(t)
+			args := map[string]any{
+				"To": GetTarget().(echo.Instances),
+			}
+			t.ConfigIstio().EvalFile(apps.Namespace.Name(), args, "testdata/gateway-api.yaml").ApplyOrFail(t)
```

Comment:
gateway creation is needed only for the gateway test.

Reply:
> gateway creation is needed only for the gateway test.

changed